File: cpp/src/quotes.cpp
#include "irm/quotes.hpp"

#include <algorithm>
#include <charconv>
#include <fstream>
#include <sstream>
#include <stdexcept>
#include <system_error>
#include <vector>

namespace irm {

namespace {

/// Locale-independent decimal parse (std::stod honours the global C locale,
/// under which "0.25" parses as 0 in e.g. de_DE).  Leading/trailing blanks
/// are tolerated; anything else in the field is an error.
double parse_double(const std::string& field, const std::string& path) {
    std::size_t b = 0, e = field.size();
    while (b < e && (field[b] == ' ' || field[b] == '\t')) ++b;
    while (e > b && (field[e - 1] == ' ' || field[e - 1] == '\t')) --e;
    double value = 0.0;
    const auto res = std::from_chars(field.data() + b, field.data() + e, value);
    if (res.ec != std::errc() || res.ptr != field.data() + e || b == e) {
        throw std::invalid_argument("bad numeric field '" + field + "' in " + path);
    }
    return value;
}

/// Split one CSV line on commas (the bundled files contain no quoting).
std::vector<std::string> split_csv(const std::string& line) {
    std::vector<std::string> out;
    std::stringstream ss(line);
    std::string field;
    while (std::getline(ss, field, ',')) {
        // Strip a trailing CR from Windows-style line endings.
        if (!field.empty() && field.back() == '\r') field.pop_back();
        out.push_back(field);
    }
    return out;
}

/// Read all non-empty data rows (header skipped) of a CSV file.
std::vector<std::vector<std::string>> read_rows(const std::string& path) {
    std::ifstream fh(path);
    if (!fh) {
        throw std::invalid_argument("cannot open CSV file: " + path);
    }
    std::vector<std::vector<std::string>> rows;
    std::string line;
    bool header = true;
    while (std::getline(fh, line)) {
        if (header) {
            header = false;
            continue;
        }
        if (line.empty() || line == "\r") continue;
        rows.push_back(split_csv(line));
    }
    return rows;
}

void sort_by_pillar(std::vector<Instrument>& out) {
    std::stable_sort(out.begin(), out.end(), [](const Instrument& a, const Instrument& b) {
        return instrument_pillar(a) < instrument_pillar(b);
    });
}

}  // namespace

std::vector<Instrument> load_curve_quotes(const std::string& path,
                                          const std::string& curve_id) {
    std::vector<Instrument> out;
    for (const auto& row : read_rows(path)) {
        if (row.size() != 5) {
            throw std::invalid_argument("malformed row in " + path);
        }
        if (row[0] != curve_id) continue;
        const std::string& kind = row[1];
        const double start = parse_double(row[2], path);
        const double maturity = parse_double(row[3], path);
        const double rate = parse_double(row[4], path);
        if (kind == "deposit") {
            out.emplace_back(Deposit(maturity, rate));
        } else if (kind == "fra") {
            out.emplace_back(FRA(start, maturity, rate));
        } else if (kind == "swap") {
            out.emplace_back(Swap(maturity, rate));
        } else {
            throw std::invalid_argument("unknown instrument type '" + kind + "' in " + path);
        }
    }
    if (out.empty()) {
        throw std::invalid_argument("no quotes found for curve_id='" + curve_id + "' in " + path);
    }
    sort_by_pillar(out);
    return out;
}

std::vector<Instrument> load_ois_quotes(const std::string& path) {
    std::vector<Instrument> out;
    for (const auto& row : read_rows(path)) {
        if (row.size() != 2) {
            throw std::invalid_argument("malformed row in " + path);
        }
        out.emplace_back(OISSwap(parse_double(row[0], path), parse_double(row[1], path)));
    }
    if (out.empty()) {
        throw std::invalid_argument("no OIS quotes in " + path);
    }
    sort_by_pillar(out);
    return out;
}

std::pair<std::vector<double>, std::vector<double>> load_zero_yields(const std::string& path) {
    std::vector<double> ts;
    std::vector<double> ys;
    for (const auto& row : read_rows(path)) {
        if (row.size() != 2) {
            throw std::invalid_argument("malformed row in " + path);
        }
        ts.push_back(parse_double(row[0], path));
        ys.push_back(parse_double(row[1], path));
    }
    if (ts.empty()) {
        throw std::invalid_argument("no zero yields in " + path);
    }
    return {ts, ys};
}

}  // namespace irm

File: rust/src/quotes.rs
//! CSV loaders for the bundled sample market data (see `../data`).
//!
//! File formats (headers included, plain ASCII, comma-separated):
//!
//! * `curve_quotes.csv`: `curve_id,instrument,start,maturity,rate` where
//!   `instrument` is `deposit`, `fra` or `swap`; `start` is used only for
//!   FRAs (0 otherwise); times in ACT/365F years, rates decimal.
//! * `ois_quotes.csv`: `maturity,rate` (OIS par swap quotes).
//! * `zero_yields.csv`: `maturity,yield` (Vasicek calibration targets).

use std::fs;
use std::path::Path;

use crate::bootstrap::Instrument;
use crate::error::{IrmError, Result};

fn read_lines(path: &Path) -> Result<Vec<String>> {
    let text = fs::read_to_string(path)
        .map_err(|e| IrmError::Data(format!("cannot read {}: {e}", path.display())))?;
    Ok(text
        .lines()
        .map(|l| l.trim().to_string())
        .filter(|l| !l.is_empty())
        .collect())
}

fn parse_f64(field: &str, path: &Path) -> Result<f64> {
    field.trim().parse::<f64>().map_err(|_| {
        IrmError::Data(format!(
            "bad numeric field {field:?} in {}",
            path.display()
        ))
    })
}

/// Load the instruments of one curve from `curve_quotes.csv`, sorted by
/// pillar time.
pub fn load_curve_quotes(path: &Path, curve_id: &str) -> Result<Vec<Instrument>> {
    let mut out: Vec<Instrument> = Vec::new();
    for line in read_lines(path)?.iter().skip(1) {
        let fields: Vec<&str> = line.split(',').collect();
        if fields.len() != 5 {
            return Err(IrmError::Data(format!(
                "expected 5 fields in {}, got line {line:?}",
                path.display()
            )));
        }
        if fields[0] != curve_id {
            continue;
        }
        let start = parse_f64(fields[2], path)?;
        let maturity = parse_f64(fields[3], path)?;
        let rate = parse_f64(fields[4], path)?;
        let ins = match fields[1] {
            "deposit" => Instrument::deposit(maturity, rate)?,
            "fra" => Instrument::fra(start, maturity, rate)?,
            "swap" => Instrument::swap(maturity, rate)?,
            other => {
                return Err(IrmError::Data(format!(
                    "unknown instrument type {other:?} in {}",
                    path.display()
                )))
            }
        };
        out.push(ins);
    }
    if out.is_empty() {
        return Err(IrmError::Data(format!(
            "no quotes found for curve_id={curve_id:?} in {}",
            path.display()
        )));
    }
    out.sort_by(|a, b| a.pillar().total_cmp(&b.pillar()));
    Ok(out)
}

/// Load OIS par swap quotes from `ois_quotes.csv`, sorted by maturity.
pub fn load_ois_quotes(path: &Path) -> Result<Vec<Instrument>> {
    let mut out: Vec<Instrument> = Vec::new();
    for line in read_lines(path)?.iter().skip(1) {
        let fields: Vec<&str> = line.split(',').collect();
        if fields.len() != 2 {
            return Err(IrmError::Data(format!(
                "expected 2 fields in {}, got line {line:?}",
                path.display()
            )));
        }
        let maturity = parse_f64(fields[0], path)?;
        let rate = parse_f64(fields[1], path)?;
        out.push(Instrument::ois_swap(maturity, rate)?);
    }
    if out.is_empty() {
        return Err(IrmError::Data(format!("no OIS quotes in {}", path.display())));
    }
    out.sort_by(|a, b| a.pillar().total_cmp(&b.pillar()));
    Ok(out)
}

/// Load `(maturities, yields)` Vasicek calibration targets from
/// `zero_yields.csv`.
pub fn load_zero_yields(path: &Path) -> Result<(Vec<f64>, Vec<f64>)> {
    let mut ts = Vec::new();
    let mut ys = Vec::new();
    for line in read_lines(path)?.iter().skip(1) {
        let fields: Vec<&str> = line.split(',').collect();
        if fields.len() != 2 {
            return Err(IrmError::Data(format!(
                "expected 2 fields in {}, got line {line:?}",
                path.display()
            )));
        }
        ts.push(parse_f64(fields[0], path)?);
        ys.push(parse_f64(fields[1], path)?);
    }
    if ts.is_empty() {
        return Err(IrmError::Data(format!("no zero yields in {}", path.display())));
    }
    Ok((ts, ys))
}
